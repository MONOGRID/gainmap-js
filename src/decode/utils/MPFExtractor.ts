type MPFExtractorOptions = {
  debug: boolean
  extractFII: boolean
  extractNonFII: boolean
}

type MPFImage = {
  MPType: number
  size: number
  dataOffset: number
  dependantImages: number
  start: number
  isFII: boolean
  end: number
}

/**
 * MPF Extractor (Multi Picture Format Extractor)
 * By Henrik S Nilsson 2019
 *
 * Extracts images stored in images based on the MPF format (found here: https://www.cipa.jp/e/std/std-sec.html
 * under "CIPA DC-007-Translation-2021 Multi-Picture Format"
 *
 * Overly commented, and without intention of being complete or production ready.
 * Created to extract depth maps from iPhone images, and to learn about image metadata.
 * Kudos to: Phil Harvey (exiftool), Jaume Sanchez (android-lens-blur-depth-extractor)
 */
export class MPFExtractor {
  options: MPFExtractorOptions

  constructor (options?: Partial<MPFExtractorOptions>) {
    this.options = {
      debug: options && options.debug !== undefined ? options.debug : false,
      extractFII: options && options.extractFII !== undefined ? options.extractFII : true,
      extractNonFII: options && options.extractNonFII !== undefined ? options.extractNonFII : true
    }
  }

  extract (imageArrayBuffer: Uint8Array) {
    return new Promise<Blob[]>((resolve, reject) => {
      const debug = this.options.debug

      const dataView = new DataView(imageArrayBuffer.buffer)
      // If you're executing this line on a big endian machine, it'll be reversed.
      // bigEnd further down though, refers to the endianness of the image itself.
      if (dataView.getUint16(0) !== 0xffd8) {
        reject('Not a valid jpeg')
        return
      }

      const length = dataView.byteLength
      let offset = 2
      let loops = 0
      let marker // APP# marker

      while (offset < length) {
        if (++loops > 250) {
          reject(`Found no marker after ${loops} loops ðŸ˜µ`)
          return
        }
        if (dataView.getUint8(offset) !== 0xff) {
          reject(`Not a valid marker at offset 0x${offset.toString(16)}, found: 0x${dataView.getUint8(offset).toString(16)}`)
          return
        }

        marker = dataView.getUint8(offset + 1)
        if (debug) console.log(`Marker: ${marker.toString(16)}`)

        if (marker === 0xe2) {
          if (debug) console.log('Found APP2 marker (0xffe2)')
          // Works for iPhone 8 Plus, X, and XSMax. Or any photos of MPF format.
          // Great way to visualize image information in html is using Exiftool. E.g.:
          // ./exiftool.exe -htmldump -wantTrailer photo.jpg > photo.html

          const formatPt = offset + 4
          /*
           *  Structure of the MP Format Identifier
           *
           *  Offset Addr.  | Code (Hex)  | Description
           *  +00             ff            Marker Prefix      <-- offset
           *  +01             e2            APP2
           *  +02             #n            APP2 Field Length
           *  +03             #n            APP2 Field Length
           *  +04             4d            'M'                <-- formatPt
           *  +05             50            'P'
           *  +06             46            'F'
           *  +07             00            NULL
           *                                                   <-- tiffOffset
           */
          if (dataView.getUint32(formatPt) === 0x4d504600) {
            // Found MPF tag, so we start dig out sub images

            const tiffOffset = formatPt + 4
            let bigEnd // Endianness from TIFF header

            // Test for TIFF validity and endianness
            // 0x4949 and 0x4D4D ('II' and 'MM') marks Little Endian and Big Endian
            if (dataView.getUint16(tiffOffset) === 0x4949) {
              bigEnd = false
            } else if (dataView.getUint16(tiffOffset) === 0x4d4d) {
              bigEnd = true
            } else {
              reject('No valid endianness marker found in TIFF header')
              return
            }

            if (dataView.getUint16(tiffOffset + 2, !bigEnd) !== 0x002a) {
              reject('Not valid TIFF data! (no 0x002A marker)')
              return
            }

            // 32 bit number stating the offset from the start of the 8 Byte MP Header
            // to MP Index IFD Least possible value is thus 8 (means 0 offset)
            const firstIFDOffset = dataView.getUint32(
              tiffOffset + 4,
              !bigEnd
            )

            if (firstIFDOffset < 0x00000008) {
              reject('Not valid TIFF data! (First offset less than 8)')
              return
            }

            // Move ahead to MP Index IFD
            // Assume we're at the first IFD, so firstIFDOffset points to
            // MP Index IFD and not MP Attributes IFD. (If we try extract from a sub image,
            // we fail silently here due to this assumption)
            // Count (2 Byte) | MP Index Fields a.k.a. MP Entries (count * 12 Byte) | Offset of Next IFD (4 Byte)
            const dirStart = tiffOffset + firstIFDOffset // Start of IFD (Image File Directory)
            const count = dataView.getUint16(dirStart, !bigEnd) // Count of MPEntries (2 Byte)

            // Extract info from MPEntries (starting after Count)
            const entriesStart = dirStart + 2
            let numberOfImages = 0
            for (
              let i = entriesStart;
              i < entriesStart + 12 * count;
              i += 12
            ) {
              // Each entry is 12 Bytes long
              // Check MP Index IFD tags, here we only take tag 0xb001 = Number of images
              if (dataView.getUint16(i, !bigEnd) === 0xb001) {
                // stored in Last 4 bytes of its 12 Byte entry.
                numberOfImages = dataView.getUint32(i + 8, !bigEnd)
              }
            }

            const nextIFDOffsetLen = 4 // 4 Byte offset field that appears after MP Index IFD tags
            const MPImageListValPt = dirStart + 2 + count * 12 + nextIFDOffsetLen
            const images: MPFImage[] = []

            for (let i = MPImageListValPt; i < MPImageListValPt + numberOfImages * 16; i += 16) {
              const image: MPFImage = {
                MPType: dataView.getUint32(i, !bigEnd),
                size: dataView.getUint32(i + 4, !bigEnd),
                // This offset is specified relative to the address of the MP Endian
                // field in the MP Header, unless the image is a First Individual Image,
                // in which case the value of the offset shall be NULL (0x00000000).
                dataOffset: dataView.getUint32(i + 8, !bigEnd),
                dependantImages: dataView.getUint32(i + 12, !bigEnd),
                start: -1,
                end: -1,
                isFII: false
              }

              if (!image.dataOffset) {
                // dataOffset is 0x00000000 for First Individual Image
                image.start = 0
                image.isFII = true
              } else {
                image.start = tiffOffset + image.dataOffset
                image.isFII = false
              }
              image.end = image.start + image.size
              images.push(image)
            }

            if (this.options.extractNonFII && images.length) {
              const bufferBlob = new Blob([dataView])
              const imgs: Blob[] = []
              for (const image of images) {
                if (image.isFII && !this.options.extractFII) {
                  continue // Skip FII
                }
                const imageBlob = bufferBlob.slice(image.start, image.end + 1, 'image/jpeg')
                // we don't need this
                // const imageUrl = URL.createObjectURL(imageBlob)
                // image.img = document.createElement('img')
                // image.img.src = imageUrl

                imgs.push(imageBlob)
              }
              resolve(imgs)
            }
          }
        }
        offset += 2 + dataView.getUint16(offset + 2)
      }
    })
  }
}
