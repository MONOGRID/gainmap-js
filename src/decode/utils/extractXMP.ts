import { GainMapMetadata } from '../../core/types'

const getAttribute = (description: Element, name: string, defaultValue?: string) => {
  let returnValue: string | [string, string, string]
  const parsedValue = description.attributes.getNamedItem(name)?.nodeValue
  if (!parsedValue) {
    const node = description.getElementsByTagName(name)[0]
    if (node) {
      const values = node.getElementsByTagName('rdf:li')
      if (values.length === 3) {
        returnValue = Array.from(values).map(v => v.innerHTML) as [string, string, string]
      } else {
        throw new Error(`Gainmap metadata contains an array of items for ${name} but its length is not 3`)
      }
    } else {
      if (defaultValue) return defaultValue
      else throw new Error(`Can't find ${name} in gainmap metadata`)
    }
  } else {
    returnValue = parsedValue
  }

  return returnValue
}

/**
 *
 * @param input
 * @returns
 */
export const extractXMP = (input: Uint8Array): GainMapMetadata | undefined => {
  let str: string
  // support node test environment
  if (typeof TextDecoder !== 'undefined') str = new TextDecoder().decode(input)
  else str = input.toString()

  let start = str.indexOf('<x:xmpmeta')
  const parser = new DOMParser()

  while (start !== -1) {
    const end = str.indexOf('x:xmpmeta>', start)
    str.slice(start, end + 10)
    const xmpBlock = str.slice(start, end + 10)
    try {
      const xmlDocument = parser.parseFromString(xmpBlock, 'text/xml')
      const description = xmlDocument.getElementsByTagName('rdf:Description')[0]

      const gainMapMin = getAttribute(description, 'hdrgm:GainMapMin', '0')
      const gainMapMax = getAttribute(description, 'hdrgm:GainMapMax')

      const gamma = getAttribute(description, 'hdrgm:Gamma', '1')

      const offsetSDR = getAttribute(description, 'hdrgm:OffsetSDR', '0.015625')
      const offsetHDR = getAttribute(description, 'hdrgm:OffsetHDR', '0.015625')

      let hdrCapacityMin = description.attributes.getNamedItem('hdrgm:HDRCapacityMin')?.nodeValue
      if (!hdrCapacityMin) hdrCapacityMin = '0'

      const hdrCapacityMax = description.attributes.getNamedItem('hdrgm:HDRCapacityMax')?.nodeValue
      if (!hdrCapacityMax) throw new Error('Incomplete gainmap metadata')

      return {
        gainMapMin: Array.isArray(gainMapMin) ? gainMapMin.map(v => parseFloat(v)) as [number, number, number] : [parseFloat(gainMapMin), parseFloat(gainMapMin), parseFloat(gainMapMin)],
        gainMapMax: Array.isArray(gainMapMax) ? gainMapMax.map(v => parseFloat(v)) as [number, number, number] : [parseFloat(gainMapMax), parseFloat(gainMapMax), parseFloat(gainMapMax)],
        gamma: Array.isArray(gamma) ? gamma.map(v => parseFloat(v)) as [number, number, number] : [parseFloat(gamma), parseFloat(gamma), parseFloat(gamma)],
        offsetSdr: Array.isArray(offsetSDR) ? offsetSDR.map(v => parseFloat(v)) as [number, number, number] : [parseFloat(offsetSDR), parseFloat(offsetSDR), parseFloat(offsetSDR)],
        offsetHdr: Array.isArray(offsetHDR) ? offsetHDR.map(v => parseFloat(v)) as [number, number, number] : [parseFloat(offsetHDR), parseFloat(offsetHDR), parseFloat(offsetHDR)],
        hdrCapacityMin: parseFloat(hdrCapacityMin),
        hdrCapacityMax: parseFloat(hdrCapacityMax)
      }
    } catch (e) {

    }
    start = str.indexOf('<x:xmpmeta', end)
  }
}
