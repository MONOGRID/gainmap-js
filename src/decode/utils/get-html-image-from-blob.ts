/**
 * private function, async get image from blob
 *
 * @param blob
 * @returns
 */
export const getHTMLImageFromBlob = (blob: Blob) => {
  return new Promise<HTMLImageElement>((resolve, reject) => {
    const img = document.createElement('img')
    img.onload = () => { resolve(img) }
    img.onerror = (e) => { reject(e) }
    img.src = URL.createObjectURL(blob)
  })
}
