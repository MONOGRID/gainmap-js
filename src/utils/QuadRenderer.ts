import {
  ByteType,
  ColorSpace,
  FloatType,
  HalfFloatType,
  IntType,
  LinearFilter,
  LinearMipMapLinearFilter,
  Material,
  Mesh,
  OrthographicCamera,
  PlaneGeometry,
  RepeatWrapping,
  RGBAFormat,
  Scene,
  ShortType,
  TextureDataType,
  UnsignedByteType,
  UnsignedIntType,
  WebGLRenderer,
  WebGLRenderTarget
} from 'three'
/**
 * @category General
 * @group General
 */
export type TextureDataTypeToBufferType<TType extends TextureDataType> =
  TType extends typeof UnsignedByteType ? Uint8ClampedArray :
  TType extends typeof HalfFloatType ? Uint16Array :
  TType extends typeof UnsignedIntType ? Uint32Array :
  TType extends typeof ByteType ? Int8Array :
  TType extends typeof ShortType ? Int16Array :
  TType extends typeof IntType ? Int32Array :
  TType extends typeof FloatType ? Float32Array :
  never

/**
 * Utility structure used for rendering a texture with a material
 *
 * @category General
 * @group General
 */
export class QuadRenderer<TType extends TextureDataType, TMaterial extends Material> {
  private _renderer: WebGLRenderer
  private _rendererIsDisposable: boolean = false
  private _material: TMaterial
  private _scene: Scene
  private _camera: OrthographicCamera
  private _quad: Mesh<PlaneGeometry>
  private _renderTarget: WebGLRenderTarget
  private _width: number
  private _height: number
  private _type: TType
  /**
   *
   * @param sourceTexture
   * @param renderer
   */
  constructor (width: number, height: number, type: TType, colorSpace: ColorSpace, material: TMaterial, renderer?:WebGLRenderer) {
    this._width = width
    this._height = height
    this._type = type
    this._material = material
    if (renderer) {
      this._renderer = renderer
    } else {
      this._renderer = QuadRenderer.instantiateRenderer()
      this._rendererIsDisposable = true
    }
    this._scene = new Scene()
    this._camera = new OrthographicCamera()
    this._camera.position.set(0, 0, 10)
    this._camera.left = -0.5
    this._camera.right = 0.5
    this._camera.top = 0.5
    this._camera.bottom = -0.5
    this._camera.updateProjectionMatrix()

    this._quad = new Mesh(new PlaneGeometry(), this._material)

    this._quad.geometry.computeBoundingBox()
    this._scene.add(this._quad)

    this._renderTarget = new WebGLRenderTarget(width, height, {
      type,
      colorSpace,
      format: RGBAFormat,
      magFilter: LinearFilter,
      minFilter: LinearMipMapLinearFilter,
      wrapS: RepeatWrapping,
      wrapT: RepeatWrapping,
      depthBuffer: false,
      stencilBuffer: false,
      generateMipmaps: true
    })
  }

  /**
   * Instantiates a temporary renderer
   *
   * @returns
   */
  public static instantiateRenderer () {
    const renderer = new WebGLRenderer()
    renderer.setSize(128, 128)
    // renderer.outputColorSpace = SRGBColorSpace
    // renderer.toneMapping = LinearToneMapping
    // renderer.debug.checkShaderErrors = false
    // this._rendererIsDisposable = true
    return renderer
  }

  /**
   * Renders the input texture using the specified material
   *
   * @param from
   * @param from
   */
  public render = () => {
    this._renderer.setRenderTarget(this._renderTarget)
    try {
      this._renderer.render(this._scene, this._camera)
    } catch (e) {
      this._renderer.setRenderTarget(null)
      throw e
    }
    this._renderer.setRenderTarget(null)
  }

  /**
   * Obtains a Buffer containing the rendered texture.
   *
   * @returns
   */
  public toArray (): TextureDataTypeToBufferType<TType> {
    let out: ArrayBufferLike
    switch (this._type) {
      case UnsignedByteType:
        out = new Uint8ClampedArray(this._width * this._height * 4)
        break
      case HalfFloatType:
        out = new Uint16Array(this._width * this._height * 4)
        break
      case UnsignedIntType:
        out = new Uint32Array(this._width * this._height * 4)
        break
      case ByteType:
        out = new Int8Array(this._width * this._height * 4)
        break
      case ShortType:
        out = new Int16Array(this._width * this._height * 4)
        break
      case IntType:
        out = new Int32Array(this._width * this._height * 4)
        break
      case FloatType:
        out = new Float32Array(this._width * this._height * 4)
        break
      default:
        throw new Error('Unsupported data type')
    }
    this._renderer.readRenderTargetPixels(this._renderTarget, 0, 0, this._width, this._height, out)
    return out as TextureDataTypeToBufferType<TType>
  }

  /**
   * If using a disposable renderer, it will dispose it.
   */
  public dispose () {
    this._renderer.setRenderTarget(null)
    if (this._rendererIsDisposable) {
      this._renderer.dispose()
      this._renderer.forceContextLoss()
    }
  }

  /**
   * Width of the texture
   */
  public get width () { return this._width }
  public set width (value: number) {
    this._width = value
    this._renderTarget.setSize(this._width, this._height)
  }

  /**
   * Height of the texture
   */
  public get height () { return this._height }
  public set height (value: number) {
    this._height = value
    this._renderTarget.setSize(this._width, this._height)
  }

  /**
   * The renderer used
   */
  public get renderer () { return this._renderer }

  /**
   * The `WebGLRenderTarget` used.
   */
  public get renderTarget () { return this._renderTarget }
  public set renderTarget (value: WebGLRenderTarget) {
    this._renderTarget = value
    this._width = value.width
    this._height = value.height
    // this._type = value.texture.type
  }

  /**
   * The `Material` used.
   */
  public get material () { return this._material }
  // public set material (value: TMaterial) { this._material = value }

  // public get rendererIsDisposable () { return this._rendererIsDisposable }
  // public set rendererIsDisposable (value: boolean) { this._rendererIsDisposable = value }
}
