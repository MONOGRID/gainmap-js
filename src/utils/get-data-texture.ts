import { DataTexture, LinearFilter, NoColorSpace, RepeatWrapping, RGBAFormat, UVMapping } from 'three'
import { EXR } from 'three/examples/jsm/loaders/EXRLoader'
import { LogLuv } from 'three/examples/jsm/loaders/LogLuvLoader'
import { RGBE } from 'three/examples/jsm/loaders/RGBELoader'
/**
 * Utility function to obtain a `DataTexture` from various input formats
 *
 * @param image
 * @returns
 */
export const getDataTexture = (image: EXR | RGBE | LogLuv | DataTexture) => {
  let dataTexture: DataTexture

  if (image instanceof DataTexture) {
    if (!(image.image.data instanceof Uint16Array) && !(image.image.data instanceof Float32Array)) {
      throw new Error('Provided image is not HDR')
    }
    dataTexture = image
  } else {
    dataTexture = new DataTexture(
      image.data,
      image.width,
      image.height,
      'format' in image ? image.format : RGBAFormat,
      image.type,
      UVMapping,
      RepeatWrapping,
      RepeatWrapping,
      LinearFilter,
      LinearFilter,
      16,
      'colorSpace' in image && image.colorSpace === 'srgb' ? image.colorSpace : NoColorSpace
    )
    dataTexture.needsUpdate = true
  }

  return dataTexture
}
