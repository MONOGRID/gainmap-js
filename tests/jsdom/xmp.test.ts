import { describe, expect, it } from '@jest/globals'
import { readFile } from 'fs/promises'
import path from 'path'
import { performance } from 'perf_hooks'

import { extractXMP } from '../../src/decode/utils/extract-metadata-from-jpeg'

describe('xmp', () => {
  it.concurrent.each([
    { fileName: '01.jpg' },
    { fileName: '02.jpg' },
    { fileName: '03.jpg' },
    { fileName: '04.jpg' },
    { fileName: '05.jpg' },
    { fileName: '06.jpg' },
    { fileName: '07.jpg' },
    { fileName: '08.jpg' },
    { fileName: '09.jpg' },
    { fileName: '10.jpg' },
    { fileName: 'pisa-4k.jpg' },
    { fileName: 'spruit_sunrise_4k.jpg' },
    { fileName: 'abandoned_bakery_16k.jpg' }

  ])('finds hdr xmp data in $fileName', async ({ fileName }) => {
    const file = await readFile(path.join(__dirname, `../fixtures/${fileName}`))
    performance.mark(`${fileName} xmp extract start`)
    const results = extractXMP(file)
    performance.mark(`${fileName} xmp extract end`)
    const measure = performance.measure(`${fileName} xmp extraction`, `${fileName} xmp extract start`, `${fileName} xmp extract end`)
    console.log(measure.name, measure.duration, 'ms')
    expect(results).not.toBeUndefined()
    expect(results).toMatchSnapshot()
  })
})
