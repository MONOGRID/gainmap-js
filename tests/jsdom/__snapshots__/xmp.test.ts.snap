// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xmp finds hdr xmp data in 01.jpg 1`] = `
{
  "gainMapMax": [
    2.474807,
    2.453772,
    2.44652,
  ],
  "gainMapMin": [
    -0.141127,
    -0.107269,
    -0.309696,
  ],
  "gamma": [
    0.250409,
    0.25,
    0.325208,
  ],
  "hdrCapacityMax": 2.8,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.183221,
    0.183221,
    0.183221,
  ],
  "offsetSdr": [
    0.183221,
    0.183221,
    0.183221,
  ],
}
`;

exports[`xmp finds hdr xmp data in 02.jpg 1`] = `
{
  "gainMapMax": [
    3.522114,
    3.060263,
    2.706481,
  ],
  "gainMapMin": [
    0.000312,
    0.034719,
    0.053824,
  ],
  "gamma": [
    0.336811,
    0.357914,
    0.386615,
  ],
  "hdrCapacityMax": 4,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.225551,
    0.225551,
    0.225551,
  ],
  "offsetSdr": [
    0.225551,
    0.225551,
    0.225551,
  ],
}
`;

exports[`xmp finds hdr xmp data in 03.jpg 1`] = `
{
  "gainMapMax": [
    0.849514,
    0.748561,
    0.753096,
  ],
  "gainMapMin": [
    0.000017,
    -0.000003,
    0.000012,
  ],
  "gamma": [
    0.25,
    0.25,
    0.25,
  ],
  "hdrCapacityMax": 4,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    10.308957,
    10.308957,
    10.308957,
  ],
  "offsetSdr": [
    10.308957,
    10.308957,
    10.308957,
  ],
}
`;

exports[`xmp finds hdr xmp data in 04.jpg 1`] = `
{
  "gainMapMax": [
    3.527605,
    2.830234,
    1.537243,
  ],
  "gainMapMin": [
    0.023869,
    0.075191,
    0.142298,
  ],
  "gamma": [
    0.506828,
    0.590032,
    1.517708,
  ],
  "hdrCapacityMax": 3.9,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.046983,
    0.046983,
    0.046983,
  ],
  "offsetSdr": [
    0.046983,
    0.046983,
    0.046983,
  ],
}
`;

exports[`xmp finds hdr xmp data in 05.jpg 1`] = `
{
  "gainMapMax": [
    1.560281,
    1.560309,
    1.560328,
  ],
  "gainMapMin": [
    -0.014377,
    -0.005596,
    -0.004277,
  ],
  "gamma": [
    0.25,
    0.25,
    0.25,
  ],
  "hdrCapacityMax": 4,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    6.695587,
    6.695587,
    6.695587,
  ],
  "offsetSdr": [
    6.695587,
    6.695587,
    6.695587,
  ],
}
`;

exports[`xmp finds hdr xmp data in 06.jpg 1`] = `
{
  "gainMapMax": [
    3.482396,
    3.126128,
    2.75434,
  ],
  "gainMapMin": [
    -0.144892,
    -0.140972,
    -0.065712,
  ],
  "gamma": [
    0.252136,
    0.263644,
    0.25,
  ],
  "hdrCapacityMax": 4,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.37841,
    0.37841,
    0.37841,
  ],
  "offsetSdr": [
    0.37841,
    0.37841,
    0.37841,
  ],
}
`;

exports[`xmp finds hdr xmp data in 07.jpg 1`] = `
{
  "gainMapMax": [
    3.961703,
    3.737166,
    3.259504,
  ],
  "gainMapMin": [
    -0.950004,
    -1.06685,
    -1.203979,
  ],
  "gamma": [
    0.622845,
    0.726034,
    0.869492,
  ],
  "hdrCapacityMax": 4,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.028746,
    0.028746,
    0.028746,
  ],
  "offsetSdr": [
    0.028746,
    0.028746,
    0.028746,
  ],
}
`;

exports[`xmp finds hdr xmp data in 08.jpg 1`] = `
{
  "gainMapMax": [
    2.668896,
    2.580987,
    2.457281,
  ],
  "gainMapMin": [
    -0.012107,
    -0.028537,
    -0.071069,
  ],
  "gamma": [
    0.260823,
    0.354576,
    0.47235,
  ],
  "hdrCapacityMax": 4,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.09302,
    0.09302,
    0.09302,
  ],
  "offsetSdr": [
    0.09302,
    0.09302,
    0.09302,
  ],
}
`;

exports[`xmp finds hdr xmp data in 09.jpg 1`] = `
{
  "gainMapMax": [
    3.270427,
    3.225892,
    3.196919,
  ],
  "gainMapMin": [
    -0.551779,
    -0.542884,
    -0.529494,
  ],
  "gamma": [
    0.42047,
    0.420672,
    0.421504,
  ],
  "hdrCapacityMax": 3.9,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.017543,
    0.017543,
    0.017543,
  ],
  "offsetSdr": [
    0.017543,
    0.017543,
    0.017543,
  ],
}
`;

exports[`xmp finds hdr xmp data in 10.jpg 1`] = `
{
  "gainMapMax": [
    2.941015,
    2.410044,
    1.595006,
  ],
  "gainMapMin": [
    -0.135098,
    -0.133564,
    -0.128712,
  ],
  "gamma": [
    0.36099,
    0.406118,
    0.538082,
  ],
  "hdrCapacityMax": 4,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.02503,
    0.02503,
    0.02503,
  ],
  "offsetSdr": [
    0.02503,
    0.02503,
    0.02503,
  ],
}
`;

exports[`xmp finds hdr xmp data in abandoned_bakery_16k.jpg 1`] = `
{
  "gainMapMax": [
    12.6294,
    12.6294,
    12.6294,
  ],
  "gainMapMin": [
    0,
    0,
    0,
  ],
  "gamma": [
    1,
    1,
    1,
  ],
  "hdrCapacityMax": 12.6294,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
  "offsetSdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
}
`;

exports[`xmp finds hdr xmp data in pisa-4k.jpg 1`] = `
{
  "gainMapMax": [
    3.13443,
    3.13443,
    3.13443,
  ],
  "gainMapMin": [
    0,
    0,
    0,
  ],
  "gamma": [
    1,
    1,
    1,
  ],
  "hdrCapacityMax": 3.13443,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
  "offsetSdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
}
`;

exports[`xmp finds hdr xmp data in spruit_sunrise_4k.jpg 1`] = `
{
  "gainMapMax": [
    15.9993,
    15.9993,
    15.9993,
  ],
  "gainMapMin": [
    0,
    0,
    0,
  ],
  "gamma": [
    1,
    1,
    1,
  ],
  "hdrCapacityMax": 15.9993,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
  "offsetSdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
}
`;
